src/core/MixHelpers.cpp

```diff
@@ -79,56 +74,12 @@ bool isSilent( const sampleFrame* src, int frames )
 	return true;
 }
 
-bool useNaNHandler()
-{
-	return s_NaNHandler;
-}
-
-void setNaNHandler( bool use )
+bool invalid(sampleFrame* buf, size_t frames)
 {
-	s_NaNHandler = use;
-}
-
-/*! \brief Function for sanitizing a buffer of infs/nans - returns true if those are found */
-bool sanitize( sampleFrame * src, int frames )
-{
-	if( !useNaNHandler() )
-	{
-		return false;
-	}
-
-	bool found = false;
-	for( int f = 0; f < frames; ++f )
-	{
-		for( int c = 0; c < 2; ++c )
-		{
-			if( std::isinf( src[f][c] ) || std::isnan( src[f][c] ) )
-			{
-				#ifdef LMMS_DEBUG
-					// TODO don't use printf here
-					printf("Bad data, clearing buffer. frame: ");
-					printf("%d: value %f\n", f, src[f][c]);
-				#endif
-				for( int f = 0; f < frames; ++f )
-				{
-					for( int c = 0; c < 2; ++c )
-					{
-						src[f][c] = 0.0f;
-					}
-				}
-				found = true;
-				return found;
-			}
-			else
-			{
-				src[f][c] = std::clamp(src[f][c], -1000.0f, 1000.0f);
-			}
-		}
-	}
-	return found;
+	const auto handler = [](auto x) { return std::isinf(x) || std::isnan(x) || x < -1.0f || x > 1.0f; };
+	return std::any_of(&buf[0][0], &buf[0][0] + frames * DEFAULT_CHANNELS, handler);
```

Comment:
It looks like you're not clipping at +/-1.0f. You clear out the buffer like with the inf/nan signals. This ain't going to work. If we clip at all I think it should be the clamp way. Also, 1.0 isn't a realistic level, at least not at this point. It's going to break projects for a lot of people who are used to cram up the level to beyond insanity.

Reply:
I think that we shouldn't consider values outside the range [-1.0, 1.0] invalid then.

Reply:
I just realized my bad misunderstanding :sob:, obviously they can be outside this range.

Reply:
It's not that they can or can't be outside of the range. That's up to the project to decide what levels are acceptable. It's setting a signal to 0 when it crosses over the set line that is undesirable.
The point of clamping at a 1000f is because it is loud enough not to interfere with artists who need to go over 1.0 and it's low enough to attenuate the shot noise from the delay effects/reverb units that come after. Usually a defunct signal will give a short peak and this will just be heard as wee tick sound. Because it will be clipped after lmms. But if there is a reverb after, the tick will loop around for a while and build up to a proper 'wall of sound'.

Reply:
> The point of clamping at a 1000f is because it is loud enough not to interfere with artists who need to go over 1.0 and it's low enough to attenuate the shot noise from the delay effects/reverb units that come after. Usually a defunct signal will give a short peak and this will just be heard as wee tick sound. Because it will be clipped after lmms. But if there is a reverb after, the tick will loop around for a while and build up to a proper 'wall of sound'.

But if it was to be over 1000.0f, the responsibility to clamp that sample down is not in this function, but before it is fed to the audio backend.

We should never be trying to mangle with the signal if its not clipping it down before it is fed to the audio backend. It wouldn't even matter anyway because `AudioEngine::clip` will clip the value that is beyond 1000.0f to just 1.0f. If one of audio devices aren't doing that, the fault is somewhere else, not here. I don't know why you would want to attenuate it much much before it is fed to the audio device.

We should fix the loud shot noise, not treat the problem by attenuating it. Users will still be dealing with a loud shot noise, just quieter. What good does that do?

Reply:
They can do so if they opt out of sanitizing the signal. It's just a trade-off. The inf/nan signals come with a lot of noise before they have reached their destination at inf/nan. Those are loud signals indeed. I don't think anyone actually amplifies their signals to beyond 1000.f.

> Users will still be dealing with a loud shot noise, just quieter. What good does that do?

It's much quieter. It's less unpleasant/scary and less likely to cause harm to speakers and neighbor relationships.

Reply:
We shouldn't do anything **ever** to the signal that the user did not ask for. If the user did not ask for the loud signal and is clearly something that resembles an issue, we fix the loud signal. That's it. We just fix it. Doing stuff like this, you never know what things you could be indirectly affecting.

> It's much quieter. It's less unpleasant/scary and less likely to cause harm to speakers and neighbor relationships.

Lets say the loud sample was at 3000.0f and the audio device clipped the signal appropriately. The audio device will clamp this down to 1.0f. If it was 90502197.0f, it will clamp it down to 1.0f. So clamping it down to 1000.0f does not do us a justice, since if it was clamped down to 1000.0f, it will end up being 1.0f anyways (assuming no further gain reduction is applied), that is, as loud as it can be. I'm not sure how it could be quieter without further gain reduction?

Reply:
To @zonkmachine, I want to apologize if I was being rude a couple of weeks ago. The main point I wanted to say was that a loud signal by itself doesn't necessarily mean that theres a problem, so we can't just clamp it to be between `[-1000, 1000]`. The point of this PR is to make the Inf/NaN issues more apparent to find, while also reworking how the detection of Inf's and NaN's work in the code. If something is causing such a loud signal before reaching the point of an `Inf` or `NaN`, that's even more reason to take a look and see what's going on.